lox_services/proxy/brightdata.py

```diff
@@ -69,122 +76,137 @@ def _excecute_request_with_retry(request_method: Callable, options: dict, retrie
-
-            ## Examples
-            ```python
-            proxy_manager = BrightDataProxyManager(brightdata_username, brightdata_password, brightdata_api_token)
-            test_ips = proxy_manager.request_with_multithreading_proxies(
-                request_method=requests.get,
-                request_options=[{"url": "https://api.ipify.org?format=json"} for _ in range(20)],
-                countries=["FR", "NL"],
-                number_of_threads=20
-            )
-            ```
+        ## Arguments
+        - `request_method`: Must be a requests method like requests.get ot requests.post or requests.put, etc...
+        - `request_options`: The list of options to used by the request_method.
+        - `countries`: A list of countries where the proxies should be.
+        - `number_of_threads`: The number of threads to use.
+        - `max_use_per_proxy`: The maximum number of use per ip for this call.
+        - `show_progress`: If True, the function will print a pretty progress bar.
+
+        ## Returns
+        - A list of requests.Response objects
+
+        ## Examples
+        ```python
+        proxy_manager = BrightDataProxyManager(brightdata_username, brightdata_password, brightdata_api_token)
+        test_ips = proxy_manager.request_with_multithreading_proxies(
+            request_method=requests.get,
+            request_options=[{"url": "https://api.ipify.org?format=json"} for _ in range(20)],
+            countries=["FR", "NL"],
+            number_of_threads=20
+        )
+        ```
         """
         number_of_threads = min(50, number_of_threads)
-        proxies = self._get_available_ips_per_countries(countries) 
-        options_with_proxies = self._populate_proxies_into_request_options(request_options, proxies, max_use_per_proxy)
-        # options_with_proxies = request_options
-        options_len = len(request_options)
-        
+        proxies = self._get_available_ips_per_countries(countries)
+        options_with_proxies = self._populate_proxies_into_request_options(
+            request_options, proxies, max_use_per_proxy
+        )
+
+        # The right way to add a progress bar with multithreading -
+        # https://stackoverflow.com/a/63834834
         results = []
+        options_len = len(options_with_proxies)
         with ThreadPoolExecutor(max_workers=number_of_threads) as executor:
-            for result in executor.map(self._excecute_request_with_retry, repeat(request_method), options_with_proxies):
-                if log and len(results) % max(1, int(options_len / 10)) == 0:
-                    print(f"{len(results)} / {options_len}")
-                
-                results.append(result)
-        
+            futures = [
+                executor.submit(
+                    self._excecute_request_with_retry, request_method, option
+                )
+                for option in options_with_proxies
+            ]
+            if show_progress:
```

Comment:
Having three if statement in a multithreading loop is very heavy process knowing how much time we execute these statements
I did a few tests, with show_progress = False, I have between 2 and 3 times more time needed for completion

Can we just duplicate the function to have one with and one without progress? The the one without progress don't have to run extra if statement? When use over 1M of threads so this is quickly a lot of computational power and time